cpp/cmake/thirdparty/get_arrow.cmake

```diff
@@ -365,7 +365,7 @@ endfunction()
 
 if(NOT DEFINED CUDF_VERSION_Arrow)
   set(CUDF_VERSION_Arrow
-      9.0.0
+      10.0.1
```

Comment:
Yup

Reply:
What is the behavior if someone has 10.0.0 installed and we try to find it? Will it reject it because we have this 10.0.1 version?

What happens if there's subsequent 10.0.2 and 10.0.3 patch releases? Would we continue to upgrade this version and potentially break being able to find and build with previously patch versions?

Reply:
AFAIK that depends on Arrow's CMake code. Looking at the version config file in my local copy, I think the answer is yes; it seems like it will automatically fail on any version strictly less than the requested version. I guess this relates to my other question, are arrow libraries ABI compatible across patch versions (hopefully)? Minor versions? Is it safe to compiler against 10.0.0 then use in a system with libarrow 10.0.1 installed?

Reply:
Yes, Arrow libraries are ABI compatible across patch versions. Arrow does not do minor releases at this time, but if it does in the future it would be ABI compatible as well. Yes it is safe to compile against `10.0.0` and then use `10.0.1`.

Reply:
In that case I think it would be sufficient to list a lower bound here (10.0.0) (or if we want to protect against Arrow 11 I think something like `10.0.0...<11.0.0`).

Reply:
Unfortunately this function isn't set up to handle that. This version feeds into the `CPMFindPackage` call in two places:
* To specify a version for the `find_package` call
  * https://github.com/rapidsai/cudf/pull/12327/files/0f3fb8bde7ffe17618aad04c3de02b04b454c8fb#diff-37169a2940bf7bc08ba411474ff544309f1baadbb3c0c3713efbb6d82c0b594fR151
* To define the git tag to use for when we want to build Arrow
  * https://github.com/rapidsai/cudf/pull/12327/files/0f3fb8bde7ffe17618aad04c3de02b04b454c8fb#diff-37169a2940bf7bc08ba411474ff544309f1baadbb3c0c3713efbb6d82c0b594fR156
  
I think we'd need to split things into two variables of something like a `find_version_spec` and `build_version`, which in our case would be `10.0.0...<11.0.0` and `10.0.1` respectively.

Alternatively, we could introduce logic to grab the list of available git tags and select the newest 10.0.x version, but I'm not sure we'd want to do that.

Reply:
Splitting on two versions would be possible. The ranges isn't an issue as we would just say `10` and that implicitly means `10.0.0...<11.0.0` since `rapids_export` would state it is only compatible across the same version value

( See explicit version entry in https://docs.rapids.ai/api/rapids-cmake/stable/command/rapids_export.html )

Reply:
Yeah assuming that the range is only ever within a single major release I agree that the `rapids_export` behavior would be equivalent to the ranges.

Reply:
My main concern with this is related to conda-forge, where libcudf 23.02 will get released with say Arrow 10.0.1, but then there's eventually an Arrow 10.0.2 and 10.0.3 and conda-forge will update its pinning and new versions of packages will end up with the dependency because currently pyarrow pins to exact builds of libarrow (this should probably be addressed as well).

Users could then end up being forced to use older versions of other packages (which they may have wanted newer versions of for new features), which is super non-ideal. Note this is only in the building a library / C++ application path and not in the runtime path as far as I can tell, so it is a bit niche, but it definitely impacts me directly :joy:.

Reply:
I am confused on the version numbers in `get_arrow` will effect conda pins / users. The `find_package` call will accept any patch of arrow 10.0 without issue.

Reply:
If we pass in `10.0.1` here won't that get passed into the `find_package` call here: https://github.com/rapidsai/cudf/pull/12327/files/0f3fb8bde7ffe17618aad04c3de02b04b454c8fb#diff-37169a2940bf7bc08ba411474ff544309f1baadbb3c0c3713efbb6d82c0b594fR151

Then if I'm reading the Arrow cmake files correctly it treats that call as effectively `>=10.0.1,<11`.

Reply:
I think I got things backwards in my head. As long as libcudf doesn't jump ahead of the conda-forge pinning I think we're fine.

Reply:
Just to clarify, your real concern would be if libcudf required `>=10.0.2,<11` while other conda-forge packages were still pinned to `<=10.0.1`? I agree, that's something we'd need to avoid.

Implementing #5097 would also simplify these discussions substantially since then the only place any of this would be maintained is in the pyarrow pinnings in conda meta.yaml files and in setup.py for pip.